gps_common/src/utm_odometry_node.cpp

```diff
@@ -11,166 +11,194 @@
-  origin_pub.publish(origin);
-
-  const Eigen::Vector3d translation(origin.pose.position.x,
-    origin.pose.position.y,
-    origin.pose.position.z);
-  const Eigen::Quaterniond rotation(origin.pose.orientation.w,
-    origin.pose.orientation.x,
-    origin.pose.orientation.y,
-    origin.pose.orientation.z);
-  world_to_local_frame = std::make_pair(rotation * -translation, rotation);
-
-  odom_pub = node.advertise<nav_msgs::Odometry>("/odom", 10);
-  world_odom_pub = node.advertise<nav_msgs::Odometry>("/odom_world", 10);
-  gps_source_sub = node.subscribe("/inspva", 10, insCallBack);
-
-  ros::spin();
-}
 
+class UTMOdometryNode {
+  private:
+    ros::NodeHandle node;
+    ros::NodeHandle priv_node;
+
+    std::string frame_id, child_frame_id, world_frame_id;
+    bool publish_odom_tf, ignore_z;
+    Rigid3d world_to_local_frame;
+
+    ros::Publisher odom_pub, world_odom_pub, origin_pub;
+    ros::Subscriber inspva_sub;
+    tf::TransformBroadcaster odom_tf_broadcaster;
+    geometry_msgs::TransformStamped odom_tf;
+    nav_msgs::Odometry odom;
+    nav_msgs::Odometry odom_world;
+
+    void InitOrigin() {
+      geometry_msgs::PoseStamped origin;
+      bool use_fixed_origin;
+      priv_node.param<bool>("use_fixed_origin", use_fixed_origin, false);
+
+      // if use_fixed_origin is false, wait for a /inspva message
+      // and use it as the origin
+      if (!use_fixed_origin) {
+        auto ins = ros::topic::waitForMessage<novatel_gps_msgs::Inspva>("/inspva", node);
+        double northing, easting;
+        std::string zone;
+        LLtoUTM(ins->latitude, ins->longitude, northing, easting, zone);
+        origin.pose.position.x = easting;
+        origin.pose.position.y = northing;
+        origin.pose.position.z = ins->height;
+        origin.pose.orientation = tf::createQuaternionMsgFromRollPitchYaw(
+          ins->roll * DEG2RAD,
+          ins->pitch * DEG2RAD,
+          (90.0 - ins->azimuth) * DEG2RAD);
+      } else {
+        // TODO more than parameters, a service call would be better
+        // so that we could change the origin at runtime
+        priv_node.param<double>("origin_x", origin.pose.position.x, 0);
+        priv_node.param<double>("origin_y", origin.pose.position.y, 0);
+        priv_node.param<double>("origin_z", origin.pose.position.z, 0);
+        // TODO using euler angles might be easier
+        priv_node.param<double>("origin_ox", origin.pose.orientation.x, 0);
+        priv_node.param<double>("origin_oy", origin.pose.orientation.y, 0);
+        priv_node.param<double>("origin_oz", origin.pose.orientation.z, 0);
+        priv_node.param<double>("origin_ow", origin.pose.orientation.w, 1);
+      }
+
+      if (ignore_z)
+        origin.pose.position.z = 0;
+
+      origin.header.frame_id = world_frame_id;
+      origin.header.stamp = ros::Time::now();
+
+      PublishOrigin(origin);
+    }
+
+    void PublishOrigin(const geometry_msgs::PoseStamped& origin) {
```

Comment:
member functions are camelCase according to google style guide.

- [ ] change to publishOrigin(const geometry_msgs::PoseStamped& origin)